README.md
oclGaussCrack
=============

Crack the verification hash of the encrypted payload of the Gauss Virus

* Name....: oclGaussCrack
* Version.: v1.4
* Autor...: Jens Steube <jens.steube@gmail.com>
* License.: GPLv2
* Language: C, OpenCL

Why
-------------

The goal of the program is to Crack the verification hash of the encrypted payload of the Gauss Virus as described here: https://www.securelist.com/en/blog/208193781/The_Mystery_of_the_Encrypted_Gauss_Payload

* Uses OpenCL to accelerate the 10k MD5 loop
* Uses optimizations also used in oclHashcat-plus for maximum performance
* Able to handle multi-GPU setups (of the same type)
* VCL (Virtual CL) v1.17 compatible
* Open Source
* Supports integration into distributed computing environments
* Supports resume

Install
-------------

Make sure you have OpenCL enabled driver installed

If you want to compile it from sources:

Make sure to have an OpenCL SDK installed
Just run "make" to compile

Performance
-------------

    AMD Radeon HD 7970              (GPU)   489k c/s
    AMD Radeon HD 6990              (GPU)   670k c/s
    AMD FX 8120                     (CPU)    14k c/s
    AMD FX 6100                     (CPU)    11k c/s
    NVIDIA GeForce GTX 470          (GPU)    94k c/s
    NVIDIA GeForce GTX 560Ti        (GPU)    86k c/s

Intel CPU supporting OpenCL should work as well.

You can increase the performance slightly by increasing the workload GPU_ACCEL in oclGaussCrack.c.
To enable CPU mode change DEV_TYPE to CL_DEVICE_TYPE_CPU.

How to use
-------------

* The program waits for any arbitrary input data on stdin. This is your password / path / the unknown key.
* It then appends the fixed salt to the input and processes the first MD5 on CPU.
* The resulting digest is used as input for the 10k MD5 loops which is done on the GPU.
* The hashes are compared on GPU. If they match, the GID which matched is stored in the result buffer.
* The host program reads the result buffer and if the hash was cracked it uses the GID to lookup the original plaintext used.

Example
-------------

This is the official Kaspersky test-vector!

NOTE: To enable "test mode", go to src/oclGaussCrack.cl and uncomment the target hash for the test vector in lines 31 - 37.

    $ ./oclGaussCrack64.bin < test.txt

This should output:

    Loading Kernel...
    Initializing OpenCL...
    Found new device # 0: Tahiti, 32 compute units
    Initialization done, accepting candidates from stdin...
    
    
    GPU # 0: ALARM! Candidate number 0 cracked the hash! Hex dump following:
    
    43 00 3a 00 5c 00 44 00
    6f 00 63 00 75 00 6d 00
    65 00 6e 00 74 00 73 00
    20 00 61 00 6e 00 64 00
    20 00 53 00 65 00 74 00
    74 00 69 00 6e 00 67 00
    73 00 5c 00 6a 00 6f 00
    68 00 6e 00 5c 00 4c 00
    6f 00 63 00 61 00 6c 00
    20 00 53 00 65 00 74 00
    74 00 69 00 6e 00 67 00
    73 00 5c 00 41 00 70 00
    70 00 6c 00 69 00 63 00
    61 00 74 00 69 00 6f 00
    6e 00 20 00 44 00 61 00
    74 00 61 00 5c 00 47 00
    6f 00 6f 00 67 00 6c 00
    65 00 5c 00 43 00 68 00
    72 00 6f 00 6d 00 65 00
    5c 00 41 00 70 00 70 00
    6c 00 69 00 63 00 61 00
    74 00 69 00 6f 00 6e 00
    7e 00 64 00 69 00 72 00
    31 00


This is how the test.txt file has been generated:

    $ echo -n "C:\Documents and Settings\john\Local Settings\Application Data\Google\Chrome\Application~dir1" | iconv -f utf8 -t utf16le > test.txt

Parameters
-------------

oclGaussCrack supports two optional parameters since v1.2.

* The first one sets an offset that defines how many entries from input data to skip.
* The second one sets an limitation that defines how many entries from input data to process.

By doing this, you can pick a range of the workload coming on stdin.
This is required to integrate oclGaussCrack into distributed computing services, like boinc or so.
It is also useful if you want to continue your work at a specific point after it stopped for whatever reason.

Both parameters are just numbers. You can set the first one without setting the second one but not vice versa.

Limitations
-------------

* Maximum input length of the string can be configured by changing MAX_LINELEN in program header.
* Maximum number of GPUs suported can be configured by changing MAX_GPU in the program header.

gaussFilter
-------------

This tool simply skips all lines from a given input which must be encoded in utf16 in case the first character value <= 0x007a.
It is useful since gauss filters all inputs from "%PROGRAMFILES%\*" where cFileName[0] > 0x007A (UNICODE 'z')
By applying this filter we can reduce the keyspace to search a lot.

Example usage: 
    
    ./gaussFilter64.bin PROGRAMFILES.dict > PROGRAMFILES_filtered.dict

gaussCombinator
-------------

This tool simply concatinates two input sources encoded in utf16 in memory.
It is useful since there are two input sources used in gauss to generate the key.

Example usage: 

    ./gaussCombinator64.bin GetEnvironmentVariableW.dict PROGRAMFILES_filtered.dict | ./gaussCrack64.bin


--

USE AT YOUR OWN RISK. This software is provided 'AS IS' without any express or implied warranty of any kind.

Copyright (c) 2012-2017, Jens Steube


doc/CHANGES
v1.4 -> v1.x

o Update Makefile to simplify compilation process
o Set OpenCL version to use before loading OpenCL headers

v1.3 -> v1.4

o Move the project to GitHub
o Update Makefile to use official Khronos OpenCL Headers
o Update Makefile to fix an issue related to _mkgmtime32
o Fix a bug where global_work_size[0] results in a wrong value
o Allow the use of deprecated OpenCL API function calls

v1.2 -> v1.3

o Added two support utilities: gaussFilter and gaussCombinator, see README for details
o Added check for \r\n LFs not just \n
o Implemented Trac #47: Full verification of all 4 known hashes
o Implemented Trac #48: Saving the cracked hashes to a "pot" file
o Fixed buf: In utf16 unicode there is a zero-byte following the \n, skip it

v1.1 -> v1.2

o Added two optional parameters to pick a range of candidates to test - useful in distributed / resume.
o Reduced required host memory
o Cleaned code to make it more readable
o Changed maximum line length from BUFSIZ to configurable MAX_LINELEN
o Changed GPU kernel to more precise checks
o Changed GPU runtime in status display
o Fixed bug: Add extra buffer space for salt and padding
o Fixed bug: VECT_SIZE was handled incorrectly
o Fixed bug: kernel source terminator byte was set incorrectly
o Fixed bug: cracked display showed more data than neccessary

v1.0 -> v1.1

o Added windows precompiled binaries
o Added dedicated 32/64 bit precompiled binaries
o Added real Gauss-Virus hash to kernel
o Test vector hash is just commented out, simply uncomment it for test
o Removed OpenSSL depency
o Fixed documentation


src/md5.c
/*
 * oclGaussCrack - OpenCL accelerated Gauss hash validator
 *
 * Copyright (C) 2012 Jens Steube <jens.steube@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#define MD5M_A 0x67452301u
#define MD5M_B 0xefcdab89u
#define MD5M_C 0x98badcfeu
#define MD5M_D 0x10325476u

#define MD5S00  7u
#define MD5S01 12u
#define MD5S02 17u
#define MD5S03 22u
#define MD5S10  5u
#define MD5S11  9u
#define MD5S12 14u
#define MD5S13 20u
#define MD5S20  4u
#define MD5S21 11u
#define MD5S22 16u
#define MD5S23 23u
#define MD5S30  6u
#define MD5S31 10u
#define MD5S32 15u
#define MD5S33 21u

#define MD5C00 0xd76aa478u
#define MD5C01 0xe8c7b756u
#define MD5C02 0x242070dbu
#define MD5C03 0xc1bdceeeu
#define MD5C04 0xf57c0fafu
#define MD5C05 0x4787c62au
#define MD5C06 0xa8304613u
#define MD5C07 0xfd469501u
#define MD5C08 0x698098d8u
#define MD5C09 0x8b44f7afu
#define MD5C0a 0xffff5bb1u
#define MD5C0b 0x895cd7beu
#define MD5C0c 0x6b901122u
#define MD5C0d 0xfd987193u
#define MD5C0e 0xa679438eu
#define MD5C0f 0x49b40821u
#define MD5C10 0xf61e2562u
#define MD5C11 0xc040b340u
#define MD5C12 0x265e5a51u
#define MD5C13 0xe9b6c7aau
#define MD5C14 0xd62f105du
#define MD5C15 0x02441453u
#define MD5C16 0xd8a1e681u
#define MD5C17 0xe7d3fbc8u
#define MD5C18 0x21e1cde6u
#define MD5C19 0xc33707d6u
#define MD5C1a 0xf4d50d87u
#define MD5C1b 0x455a14edu
#define MD5C1c 0xa9e3e905u
#define MD5C1d 0xfcefa3f8u
#define MD5C1e 0x676f02d9u
#define MD5C1f 0x8d2a4c8au
#define MD5C20 0xfffa3942u
#define MD5C21 0x8771f681u
#define MD5C22 0x6d9d6122u
#define MD5C23 0xfde5380cu
#define MD5C24 0xa4beea44u
#define MD5C25 0x4bdecfa9u
#define MD5C26 0xf6bb4b60u
#define MD5C27 0xbebfbc70u
#define MD5C28 0x289b7ec6u
#define MD5C29 0xeaa127fau
#define MD5C2a 0xd4ef3085u
#define MD5C2b 0x04881d05u
#define MD5C2c 0xd9d4d039u
#define MD5C2d 0xe6db99e5u
#define MD5C2e 0x1fa27cf8u
#define MD5C2f 0xc4ac5665u
#define MD5C30 0xf4292244u
#define MD5C31 0x432aff97u
#define MD5C32 0xab9423a7u
#define MD5C33 0xfc93a039u
#define MD5C34 0x655b59c3u
#define MD5C35 0x8f0ccc92u
#define MD5C36 0xffeff47du
#define MD5C37 0x85845dd1u
#define MD5C38 0x6fa87e4fu
#define MD5C39 0xfe2ce6e0u
#define MD5C3a 0xa3014314u
#define MD5C3b 0x4e0811a1u
#define MD5C3c 0xf7537e82u
#define MD5C3d 0xbd3af235u
#define MD5C3e 0x2ad7d2bbu
#define MD5C3f 0xeb86d391u

#define MD5_F(x,y,z)    z ^ (x & (y ^ z))
#define MD5_G(x,y,z)    y ^ (z & (x ^ y))
#define MD5_H1(x,y,z)   (tmp2 = (x ^ y)) ^ z
#define MD5_H2(x,y,z)   x ^ tmp2
#define MD5_I(x,y,z)    y ^ (x | ~z)

#define MD5_Fo(x,y,z)   bitselect (z, y, x)
#define MD5_Go(x,y,z)   bitselect (y, x, z)

#define ROTL(a,n)       rotate (a, n)

#define MD5_STEP(f,a,b,c,d,x,K,s)   \
{                                   \
  a += K;                           \
  a += x;                           \
  a += f (b, c, d);                 \
  a  = ROTL (a, s);                 \
  a += b;                           \
}

#define MD5_STEP0(f,a,b,c,d,K,s)    \
{                                   \
  a += K;                           \
  a += f (b, c, d);                 \
  a  = ROTL (a, s);                 \
  a += b;                           \
}

#define MD5_STEPZ(f,a,b,c,d,x,K,s)  \
{                                   \
  a += x;                           \
  a  = ROTL (a, s);                 \
  a += b;                           \
}



src/oclGaussCrack.c
/*
 * oclGaussCrack - OpenCL accelerated Gauss hash validator
 *
 * Copyright (C) 2012 Jens Steube <jens.steube@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#define CL_TARGET_OPENCL_VERSION 120

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <search.h>
#include <errno.h>
#include <getopt.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <CL/cl.h>

#include "md5.c"
#include "OpenCL.c"

#define KERNEL_SRC    "src/oclGaussCrack.cl"
#define KERNEL_NAME   "oclGaussCrack"
#define POTFILE       "oclGaussCrack.pot"
#define BUILD_OPTS    "-I . -I src"
#define DEV_TYPE      CL_DEVICE_TYPE_GPU
#define MAX_LINELEN   256 + 100
#define MAX_PLATFORM  16
#define MAX_GPU       16
#define VECT_SIZE     4   // must be power of 2
#define GPU_THREADS   64  // must be power of 2
#define GPU_ACCEL     32

#define rotate(x,n)   (x << n) | (x >> (32 - n))

#define timer_set(a)  gettimeofday (a, NULL)

#define timer_get(a,r)                        \
{                                             \
  struct timeval hr_tmp;                      \
  timer_set (&hr_tmp);                        \
  r = ((hr_tmp.tv_sec  - a.tv_sec)  * 1000)   \
    + ((hr_tmp.tv_usec - a.tv_usec) / 1000);  \
}

typedef struct __block
{
  /* it is using uint4 in kernel thats why this looks a bit weird */

  cl_uint A[4];
  cl_uint B[4];
  cl_uint C[4];
  cl_uint D[4];

} block_t;

typedef struct
{
  cl_context          context;
  cl_program          program;
  cl_kernel           kernel;
  cl_command_queue    command_queue;

  cl_uint             max_compute_units;
  uint32_t            num_threads;
  uint32_t            num_elements;
  uint32_t            num_cached;
  uint32_t            num_work;

  uint8_t           **plains_buf;
  size_t             *plains_len;

  cl_mem              d_block;
  cl_mem              d_results;
  block_t            *h_block;
  uint32_t           *h_results;

  struct timeval      timer;

} gpu_ctx_t;

static void dump_hex (FILE *fd, const uint8_t *s, const size_t sz)
{
  for (size_t i = 0; i < sz; i += 8)
  {
    for (size_t j = 0; j < 8; j += 1)
    {
      const size_t pos = i + j;

      if (pos == sz) break;

      fprintf (fd, "%02x ", s[pos]);
    }

    fprintf (fd, "\n");
  }

  fprintf (fd, "\n");
}

static void calc_work (const cl_uint num_devices, gpu_ctx_t *gpu_ctxs)
{
  for (cl_uint device_id = 0; device_id < num_devices; device_id++)
  {
    gpu_ctx_t *gpu_ctx = &gpu_ctxs[device_id];

    if (gpu_ctx->num_cached == 0) continue;

    gpu_ctx->num_work = ((gpu_ctx->num_cached + (VECT_SIZE - 1)) & ~(VECT_SIZE - 1)) / VECT_SIZE;
  }
}

static void launch_kernel (const cl_uint num_devices, gpu_ctx_t *gpu_ctxs)
{
  for (cl_uint device_id = 0; device_id < num_devices; device_id++)
  {
    gpu_ctx_t *gpu_ctx = &gpu_ctxs[device_id];

    if (gpu_ctx->num_work == 0) continue;

    timer_set (&gpu_ctx->timer);

    const size_t size_block = gpu_ctx->num_work * sizeof (block_t);

    gc_clEnqueueWriteBuffer (gpu_ctx->command_queue, gpu_ctx->d_block, CL_FALSE, 0, size_block, gpu_ctx->h_block, 0, NULL, NULL);

    size_t global_work_size[3] = { gpu_ctx->num_work,    1, 1 };
    size_t local_work_size[3]  = { gpu_ctx->num_threads, 1, 1 };

    while (global_work_size[0] % local_work_size[0]) global_work_size[0]++;

    gc_clEnqueueNDRangeKernel (gpu_ctx->command_queue, gpu_ctx->kernel, 1, NULL, global_work_size, local_work_size, 0, NULL, NULL);

    gc_clFlush (gpu_ctx->command_queue);
  }
}

static void check_results (const cl_uint num_devices, gpu_ctx_t *gpu_ctxs)
{
  static uint64_t total = 0;

  for (cl_uint device_id = 0; device_id < num_devices; device_id++)
  {
    gpu_ctx_t *gpu_ctx = &gpu_ctxs[device_id];

    if (gpu_ctx->num_cached == 0) continue;

    const size_t size_results = gpu_ctx->num_threads * sizeof (uint32_t);

    gc_clEnqueueReadBuffer (gpu_ctx->command_queue, gpu_ctx->d_results, CL_TRUE, 0, size_results, gpu_ctx->h_results, 0, NULL, NULL);

    for (uint32_t thread = 0; thread < gpu_ctx->num_threads; thread++)
    {
      if (gpu_ctx->h_results[thread] != 0xffffffff)
      {
        uint32_t gid = gpu_ctx->h_results[thread];

        fprintf (stdout, "\nGPU #%2d: ALARM! Candidate number %u cracked the hash! Hex dump following:\n\n", device_id, gid);

        dump_hex (stdout, gpu_ctx->plains_buf[gid], gpu_ctx->plains_len[gid]);

        FILE *fd = fopen (POTFILE, "ab");

        if (fd) // ignore error
        {
          fprintf (fd, "\nGPU #%2d: ALARM! Candidate number %u cracked the hash! Hex dump following:\n\n", device_id, gid);

          dump_hex (fd, gpu_ctx->plains_buf[gid], gpu_ctx->plains_len[gid]);

          fclose (fd);
        }

        exit (0);
      }
    }

    const uint32_t num_cached = gpu_ctx->num_cached;

    total += num_cached;

    float ms;

    timer_get (gpu_ctx->timer, ms);

    if (ms == 0) continue;

    float speed = (num_cached / ms) * 1000;

    printf ("GPU #%2d: %u candidates in %u ms [%u/s] - Total = %llu...\n", device_id, num_cached, (uint32_t) ms, (uint32_t) speed, (long long unsigned int) total);
  }
}

static void md5_transform (uint32_t *W, const uint32_t len, uint32_t digest[4])
{
  // prepare some stuff

  digest[0] = MD5M_A;
  digest[1] = MD5M_B;
  digest[2] = MD5M_C;
  digest[3] = MD5M_D;

  uint32_t len_pad = (len + 0x3f) & ~0x3f;

  uint8_t *ptr = (uint8_t *) W;

  memset (ptr + len, 0, len_pad - len);

  ptr[len] = 0x80;

  W[((len_pad - 64) / 4) + 14] = len * 8;

  // loop that

  for (uint32_t i = len_pad; i >= 55; i -= 64, W += 16)
  {
    uint32_t a = digest[0];
    uint32_t b = digest[1];
    uint32_t c = digest[2];
    uint32_t d = digest[3];

    uint32_t tmp2;

    MD5_STEP (MD5_F , a, b, c, d, W[ 0], MD5C00, MD5S00);
    MD5_STEP (MD5_F , d, a, b, c, W[ 1], MD5C01, MD5S01);
    MD5_STEP (MD5_F , c, d, a, b, W[ 2], MD5C02, MD5S02);
    MD5_STEP (MD5_F , b, c, d, a, W[ 3], MD5C03, MD5S03);
    MD5_STEP (MD5_F , a, b, c, d, W[ 4], MD5C04, MD5S00);
    MD5_STEP (MD5_F , d, a, b, c, W[ 5], MD5C05, MD5S01);
    MD5_STEP (MD5_F , c, d, a, b, W[ 6], MD5C06, MD5S02);
    MD5_STEP (MD5_F , b, c, d, a, W[ 7], MD5C07, MD5S03);
    MD5_STEP (MD5_F , a, b, c, d, W[ 8], MD5C08, MD5S00);
    MD5_STEP (MD5_F , d, a, b, c, W[ 9], MD5C09, MD5S01);
    MD5_STEP (MD5_F , c, d, a, b, W[10], MD5C0a, MD5S02);
    MD5_STEP (MD5_F , b, c, d, a, W[11], MD5C0b, MD5S03);
    MD5_STEP (MD5_F , a, b, c, d, W[12], MD5C0c, MD5S00);
    MD5_STEP (MD5_F , d, a, b, c, W[13], MD5C0d, MD5S01);
    MD5_STEP (MD5_F , c, d, a, b, W[14], MD5C0e, MD5S02);
    MD5_STEP (MD5_F , b, c, d, a, W[15], MD5C0f, MD5S03);

    MD5_STEP (MD5_G , a, b, c, d, W[ 1], MD5C10, MD5S10);
    MD5_STEP (MD5_G , d, a, b, c, W[ 6], MD5C11, MD5S11);
    MD5_STEP (MD5_G , c, d, a, b, W[11], MD5C12, MD5S12);
    MD5_STEP (MD5_G , b, c, d, a, W[ 0], MD5C13, MD5S13);
    MD5_STEP (MD5_G , a, b, c, d, W[ 5], MD5C14, MD5S10);
    MD5_STEP (MD5_G , d, a, b, c, W[10], MD5C15, MD5S11);
    MD5_STEP (MD5_G , c, d, a, b, W[15], MD5C16, MD5S12);
    MD5_STEP (MD5_G , b, c, d, a, W[ 4], MD5C17, MD5S13);
    MD5_STEP (MD5_G , a, b, c, d, W[ 9], MD5C18, MD5S10);
    MD5_STEP (MD5_G , d, a, b, c, W[14], MD5C19, MD5S11);
    MD5_STEP (MD5_G , c, d, a, b, W[ 3], MD5C1a, MD5S12);
    MD5_STEP (MD5_G , b, c, d, a, W[ 8], MD5C1b, MD5S13);
    MD5_STEP (MD5_G , a, b, c, d, W[13], MD5C1c, MD5S10);
    MD5_STEP (MD5_G , d, a, b, c, W[ 2], MD5C1d, MD5S11);
    MD5_STEP (MD5_G , c, d, a, b, W[ 7], MD5C1e, MD5S12);
    MD5_STEP (MD5_G , b, c, d, a, W[12], MD5C1f, MD5S13);

    MD5_STEP (MD5_H1, a, b, c, d, W[ 5], MD5C20, MD5S20);
    MD5_STEP (MD5_H2, d, a, b, c, W[ 8], MD5C21, MD5S21);
    MD5_STEP (MD5_H1, c, d, a, b, W[11], MD5C22, MD5S22);
    MD5_STEP (MD5_H2, b, c, d, a, W[14], MD5C23, MD5S23);
    MD5_STEP (MD5_H1, a, b, c, d, W[ 1], MD5C24, MD5S20);
    MD5_STEP (MD5_H2, d, a, b, c, W[ 4], MD5C25, MD5S21);
    MD5_STEP (MD5_H1, c, d, a, b, W[ 7], MD5C26, MD5S22);
    MD5_STEP (MD5_H2, b, c, d, a, W[10], MD5C27, MD5S23);
    MD5_STEP (MD5_H1, a, b, c, d, W[13], MD5C28, MD5S20);
    MD5_STEP (MD5_H2, d, a, b, c, W[ 0], MD5C29, MD5S21);
    MD5_STEP (MD5_H1, c, d, a, b, W[ 3], MD5C2a, MD5S22);
    MD5_STEP (MD5_H2, b, c, d, a, W[ 6], MD5C2b, MD5S23);
    MD5_STEP (MD5_H1, a, b, c, d, W[ 9], MD5C2c, MD5S20);
    MD5_STEP (MD5_H2, d, a, b, c, W[12], MD5C2d, MD5S21);
    MD5_STEP (MD5_H1, c, d, a, b, W[15], MD5C2e, MD5S22);
    MD5_STEP (MD5_H2, b, c, d, a, W[ 2], MD5C2f, MD5S23);

    MD5_STEP (MD5_I , a, b, c, d, W[ 0], MD5C30, MD5S30);
    MD5_STEP (MD5_I , d, a, b, c, W[ 7], MD5C31, MD5S31);
    MD5_STEP (MD5_I , c, d, a, b, W[14], MD5C32, MD5S32);
    MD5_STEP (MD5_I , b, c, d, a, W[ 5], MD5C33, MD5S33);
    MD5_STEP (MD5_I , a, b, c, d, W[12], MD5C34, MD5S30);
    MD5_STEP (MD5_I , d, a, b, c, W[ 3], MD5C35, MD5S31);
    MD5_STEP (MD5_I , c, d, a, b, W[10], MD5C36, MD5S32);
    MD5_STEP (MD5_I , b, c, d, a, W[ 1], MD5C37, MD5S33);
    MD5_STEP (MD5_I , a, b, c, d, W[ 8], MD5C38, MD5S30);
    MD5_STEP (MD5_I , d, a, b, c, W[15], MD5C39, MD5S31);
    MD5_STEP (MD5_I , c, d, a, b, W[ 6], MD5C3a, MD5S32);
    MD5_STEP (MD5_I , b, c, d, a, W[13], MD5C3b, MD5S33);
    MD5_STEP (MD5_I , a, b, c, d, W[ 4], MD5C3c, MD5S30);
    MD5_STEP (MD5_I , d, a, b, c, W[11], MD5C3d, MD5S31);
    MD5_STEP (MD5_I , c, d, a, b, W[ 2], MD5C3e, MD5S32);
    MD5_STEP (MD5_I , b, c, d, a, W[ 9], MD5C3f, MD5S33);

    digest[0] += a;
    digest[1] += b;
    digest[2] += c;
    digest[3] += d;
  }
}

int main (int argc, char *argv[])
{
  uint64_t skip =  0;
  uint64_t left = -1;

  if (argc >= 2) skip = atoll (argv[1]);
  if (argc >= 3) left = atoll (argv[2]);

  printf ("Loading Kernel...\n");

  const char *filename = KERNEL_SRC;

  struct stat s;

  if (stat (filename, &s) == -1)
  {
    fprintf (stderr, "%s: %s in line %d\n", filename, strerror (errno), __LINE__);

    return (-1);
  }

  FILE *fp = fopen (filename, "rb");

  if (fp == NULL)
  {
    fprintf (stderr, "%s: %s in line %d\n", filename, strerror (errno), __LINE__);

    return (-1);
  }

  char *source_buf = (char *) malloc (s.st_size + 1);

  if (!fread (source_buf, sizeof (char), s.st_size, fp))
  {
    fprintf (stderr, "%s: %s in line %d\n", filename, strerror (errno), __LINE__);

    return (-1);
  }

  source_buf[s.st_size] = 0;

  fclose (fp);

  const char *sourceBuf[] = { source_buf };

  const size_t sourceLen[] = { s.st_size + 1 };

  printf ("Initializing OpenCL...\n");

  cl_platform_id platform;

  cl_uint num_devices = 0;

  cl_device_id devices[MAX_PLATFORM];

  gc_clGetPlatformIDs (1, &platform, NULL);

  gc_clGetDeviceIDs (platform, DEV_TYPE, MAX_PLATFORM, devices, &num_devices);

  gpu_ctx_t gpu_ctxs[MAX_GPU];

  memset (gpu_ctxs, 0, sizeof (gpu_ctxs));

  for (cl_uint device_id = 0; device_id < num_devices; device_id++)
  {
    cl_device_id device = devices[device_id];

    cl_context context = gc_clCreateContext (NULL, 1, &device, NULL, NULL);

    cl_program program = gc_clCreateProgramWithSource (context, 1, sourceBuf, sourceLen);

    gc_clBuildProgram (program, 1, &device, BUILD_OPTS, NULL, NULL);

    cl_kernel kernel = gc_clCreateKernel (program, KERNEL_NAME);

    cl_command_queue command_queue = gc_clCreateCommandQueue (context, device, 0);

    cl_uint max_compute_units;

    gc_clGetDeviceInfo (device, CL_DEVICE_MAX_COMPUTE_UNITS, sizeof (max_compute_units), &max_compute_units, NULL);

    char device_name[BUFSIZ];

    memset (device_name, 0, sizeof (device_name));

    gc_clGetDeviceInfo (device, CL_DEVICE_NAME, sizeof (device_name), &device_name, NULL);

    printf ("Found new device #%2d: %s, %u compute units\n", device_id, device_name, max_compute_units);

    const int num_threads  = GPU_THREADS;
    const int num_elements = max_compute_units * num_threads * GPU_ACCEL;

    /**
     * GPU memory
     */

    const size_t size_block   = num_elements * sizeof (block_t);
    const size_t size_results = num_threads  * sizeof (uint32_t);

    cl_mem d_block = gc_clCreateBuffer (context, CL_MEM_READ_ONLY, size_block, NULL);

    cl_mem d_results = gc_clCreateBuffer (context, CL_MEM_WRITE_ONLY, size_results, NULL);

    gc_clSetKernelArg (kernel, 0, sizeof (cl_mem), (void *) &d_block);
    gc_clSetKernelArg (kernel, 1, sizeof (cl_mem), (void *) &d_results);

    /**
     * Host memory
     */

    block_t *h_block = (block_t *) malloc (size_block);

    uint32_t *h_results = (uint32_t *) malloc (size_results);

    memset (h_results, 0xff, size_results);

    gc_clEnqueueWriteBuffer (command_queue, d_results, CL_TRUE, 0, size_results, h_results, 0, NULL, NULL);

    /**
     * Buffers for candidates
     */

    uint8_t **plains_buf = (uint8_t **) calloc (num_elements * VECT_SIZE, sizeof (uint8_t *));

    for (int i = 0; i < num_elements * VECT_SIZE; i++)
    {
      /* Agreed, this is not nice. But who cares nowadays? */

      plains_buf[i] = (uint8_t *) malloc (MAX_LINELEN);
    }

    size_t *plains_len = (size_t *) calloc (num_elements * VECT_SIZE, sizeof (size_t));

    gpu_ctx_t *gpu_ctx = &gpu_ctxs[device_id];

    gpu_ctx->context           = context;
    gpu_ctx->program           = program;
    gpu_ctx->kernel            = kernel;
    gpu_ctx->command_queue     = command_queue;
    gpu_ctx->max_compute_units = max_compute_units;
    gpu_ctx->d_block           = d_block;
    gpu_ctx->d_results         = d_results;
    gpu_ctx->h_block           = h_block;
    gpu_ctx->h_results         = h_results;
    gpu_ctx->num_threads       = num_threads;
    gpu_ctx->num_elements      = num_elements;
    gpu_ctx->plains_buf        = plains_buf;
    gpu_ctx->plains_len        = plains_len;
  }

  /* static salt */

  const uint8_t salt_buf[16] =
  {
    0x97, 0x48, 0x6C, 0xAA,
    0x22, 0x5F, 0xE8, 0x77,
    0xC0, 0x35, 0xCC, 0x03,
    0x73, 0x23, 0x6D, 0x51
  };

  const size_t salt_len = sizeof (salt_buf);

  /* main loop */

  printf ("Initialization done, accepting candidates from stdin...\n\n");

  cl_uint cur_device_id = 0;

  while (!feof (stdin))
  {
    /* Get new password candidate from stdin */

    uint8_t line_buf[MAX_LINELEN];

    int cur_c = 0;

    int prev_c = 0;

    size_t line_len = 0;

    for (size_t i = 0; i < MAX_LINELEN - 100; i++) // - 100 = we need some space for salt and padding
    {
      cur_c = getchar ();

      if (cur_c == EOF) break;

      if ((prev_c == '\n') && (cur_c == '\0'))
      {
        line_len--;

        break;
      }

      line_buf[line_len] = cur_c;

      line_len++;

      prev_c = cur_c;
    }

    /* chop \r if it exists for some reason (in case user used a dictionary) */

    if (line_len >= 2)
    {
      if ((prev_c == '\r') && (cur_c == '\0')) line_len -= 2;
    }

    /* skip empty lines */

    if (line_len == 0) continue;

    /* The following enables distributed computing / resume work */

    if (skip)
    {
      skip--;

      continue;
    }

    if (left)
    {
      left--;
    }
    else
    {
      break;
    }

    /* Append constant salt */

    memcpy (line_buf + line_len, salt_buf, salt_len);

    line_len += salt_len;

    /* Generate digest out of it */

    uint32_t digest[4];

    md5_transform ((uint32_t *) line_buf, (uint32_t) line_len, digest);

    /* Next garanteed free GPU */

    gpu_ctx_t *gpu_ctx = &gpu_ctxs[cur_device_id];

    /* Save original buffer in case it cracks it */

    memcpy (gpu_ctx->plains_buf[gpu_ctx->num_cached], line_buf, line_len - salt_len);

    gpu_ctx->plains_len[gpu_ctx->num_cached] = line_len - salt_len;

    /* Next garanteed free memory element on that GPU */

    const uint32_t element_div = gpu_ctx->num_cached / 4;
    const uint32_t element_mod = gpu_ctx->num_cached % 4;

    /* Copy new digest */

    gpu_ctx->h_block[element_div].A[element_mod] = digest[0];
    gpu_ctx->h_block[element_div].B[element_mod] = digest[1];
    gpu_ctx->h_block[element_div].C[element_mod] = digest[2];
    gpu_ctx->h_block[element_div].D[element_mod] = digest[3];

    gpu_ctx->num_cached++;

    /* If memory elements on that GPU are full, switch to the next GPU */

    if ((gpu_ctx->num_cached / VECT_SIZE) < gpu_ctx->num_elements) continue;

    cur_device_id++;

    /* If there is no more GPU left, run the calculation */

    if (cur_device_id < num_devices) continue;

    /* Fire! */

    calc_work (num_devices, gpu_ctxs);

    launch_kernel (num_devices, gpu_ctxs);

    /* Collecting data has a blocking effect */

    check_results (num_devices, gpu_ctxs);

    /* Reset buffer state */

    for (cl_uint device_id = 0; device_id < num_devices; device_id++)
    {
      gpu_ctx_t *gpu_ctx = &gpu_ctxs[device_id];

      gpu_ctx->num_cached = 0;
    }

    cur_device_id = 0;
  }

  /* Final calculation of leftovers */

  calc_work (num_devices, gpu_ctxs);

  launch_kernel (num_devices, gpu_ctxs);

  check_results (num_devices, gpu_ctxs);

  return -1;
}
